packages/kbn-config/src/deprecation/apply_deprecations.ts

```diff
@@ -22,22 +26,31 @@ export const applyDeprecations = (
   config: Record<string, any>,
   deprecations: ConfigDeprecationWithContext[],
   createAddDeprecation: (pluginId: string) => AddConfigDeprecation = noopAddDeprecationFactory
-) => {
+): { config: Record<string, any>; changedPaths: ChangedDeprecatedPaths } => {
   const result = cloneDeep(config);
+  const changedPaths: ChangedDeprecatedPaths = {
+    set: [],
+    unset: [],
+  };
   deprecations.forEach(({ deprecation, path }) => {
     const commands = deprecation(result, path, createAddDeprecation(path));
     if (commands) {
       if (commands.set) {
+        changedPaths.set.push(...commands.set.map((c) => c.path));
         commands.set.forEach(function ({ path: commandPath, value }) {
           set(result, commandPath, value);
```

Comment:
I think we should use `config` here instead of `result` to render mutating the configs useless inside the deprecation factory.

Alternatively we can `deepFreeze` the object that we pass instead of `cloneDeep` and then `deepMerge` the changes into a new object.